starboard/nplb/posix_compliance/posix_epoll_test.cc

```diff
@@ -0,0 +1,279 @@
+  EXPECT_EQ(epoll_ctl(epfd, EPOLL_CTL_DEL, socket_fd, &epev), 0)
+      << strerror(errno);
+
+  // Remove after remove should fail.
+  EXPECT_LT(epoll_ctl(epfd, EPOLL_CTL_DEL, socket_fd, &epev), 0);
+
+  // Add after remove should succeed.
+  EXPECT_EQ(epoll_ctl(epfd, EPOLL_CTL_ADD, socket_fd, &epev), 0)
+      << strerror(errno);
+
+  // Second add after remove should fail.
+  EXPECT_LT(epoll_ctl(epfd, EPOLL_CTL_ADD, socket_fd, &epev), 0);
+
+  // Remove should succeed, to clear epoll queue for next test.
+  EXPECT_EQ(epoll_ctl(epfd, EPOLL_CTL_DEL, socket_fd, &epev), 0)
+      << strerror(errno);
+}
+
+TEST_F(PosixEpollTest, RainyDayInvalidSocket) {
+  EXPECT_LT(epoll_ctl(epfd, EPOLL_CTL_ADD, -1, &epev), 0);
+}
+
+TEST_F(PosixEpollTest, RainyDayNoInterest) {
+  EXPECT_LT(epoll_ctl(epfd, EPOLL_CTL_ADD, socket_fd, nullptr), 0);
+}
+
+TEST_F(PosixEpollTest, SunnyDayWaitTimeout) {
+  // Create listen socket, bind and listen on <port>.
+  int listen_sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+  ASSERT_GE(listen_sock, 0);
+
+  // Set socket to be reusable.
+  const int on = 1;
+  int result =
+      setsockopt(listen_sock, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
+  EXPECT_EQ(result, 0) << strerror(errno);
+
+  // Bind socket with local address.
+  result =
+      bind(listen_sock, reinterpret_cast<struct sockaddr*>(&address), add_len);
+  EXPECT_EQ(result, 0) << strerror(errno);
+
+  result = listen(listen_sock, kMaxConn);
+  EXPECT_EQ(result, 0) << strerror(errno);
+
+  struct epoll_event ev, events[kMaxEvents];
+  ev.events = EPOLLIN;
+  ev.data.fd = listen_sock;
+  EXPECT_EQ(epoll_ctl(epfd, EPOLL_CTL_ADD, listen_sock, &ev), 0)
+      << strerror(errno);
+
+  // 2 second timeout for wait.
+  int num_ready = epoll_wait(epfd, events, kMaxEvents, kSocketTimeout / 100);
+  // Expect nothing to be ready since we're not doing anything.
+  EXPECT_EQ(num_ready, 0);
+
+  EXPECT_EQ(close(listen_sock), 0) << strerror(errno);
+}
+
+TEST_F(PosixEpollTest, ReceiveWithWait) {
+  const int kBufSize = 256 * 1024;
+  const int kSockBufSize = kBufSize / 8;
+  int listen_socket_fd = -1, client_socket_fd = -1, server_socket_fd = -1;
+  int result = PosixSocketCreateAndConnect(
+      AF_INET, AF_INET, htons(PosixGetPortNumberForTests()), kSocketTimeout,
+      &listen_socket_fd, &client_socket_fd, &server_socket_fd);
+
+  // Let's set the buffers small to create partial reads and writes.
+  PosixSocketSetReceiveBufferSize(client_socket_fd, kSockBufSize);
+  PosixSocketSetReceiveBufferSize(server_socket_fd, kSockBufSize);
+  PosixSocketSetSendBufferSize(client_socket_fd, kSockBufSize);
+  PosixSocketSetSendBufferSize(server_socket_fd, kSockBufSize);
+
+  // Create the buffers and fill the send buffer with a pattern, the receive
+  // buffer with zeros.
+  char* send_buf = new char[kBufSize];
+  char* receive_buf = new char[kBufSize];
+  for (int i = 0; i < kBufSize; ++i) {
+    send_buf[i] = static_cast<char>(i);
+    receive_buf[i] = 0;
+  }
```

Comment:
I feel a strong sense of having commented before about using std::array<> and STL functions, and either you forget those comments or are actively ignoring them.

_If_ the idea is not to use C++ traits (big IF), then at least make that plain
by filing a bug and leaving here and in other parts where these C90 (or C11) 
idioms are used (e.g. l.82, l.90) a `// TODO(b/...): Migrate to C++.`.